File: app/src/main/java/com/pavelprymak/propodcast/presentation/paging/SearchDataSourceFactory.java
package com.pavelprymak.propodcast.presentation.paging;

import androidx.annotation.NonNull;
import androidx.paging.DataSource;

import com.pavelprymak.propodcast.model.network.pojo.search.ResultsItem;
import com.pavelprymak.propodcast.model.network.repo.PodcastRepoRx;
import com.pavelprymak.propodcast.presentation.common.PagingStateBatch;

import java.util.List;

public class SearchDataSourceFactory extends DataSource.Factory<Integer, ResultsItem> {
    private final PagingStateBatch mPagingStateBatch;
    private final PodcastRepoRx mRepoRx;
    private final String mSearchQuery;
    private final String mLanguage;
    private List<ResultsItem> mPrevLoadingList;

    public SearchDataSourceFactory(PagingStateBatch pagingStateBatch, PodcastRepoRx repoRx, String searchQuery, String language) {
        mPagingStateBatch = pagingStateBatch;
        mRepoRx = repoRx;
        mSearchQuery = searchQuery;
        mLanguage = language;
    }

    public SearchDataSourceFactory(PagingStateBatch pagingStateBatch, PodcastRepoRx repoRx, String searchQuery, String language, List<ResultsItem> prevLoadingList) {
        mPagingStateBatch = pagingStateBatch;
        mRepoRx = repoRx;
        mSearchQuery = searchQuery;
        mLanguage = language;
        mPrevLoadingList = prevLoadingList;
    }

    @NonNull
    @Override
    public DataSource<Integer, ResultsItem> create() {
        return new SearchDataSource(mRepoRx, mPagingStateBatch, mSearchQuery, mLanguage, mPrevLoadingList);
    }
}

File: app/src/main/java/com/pavelprymak/propodcast/presentation/paging/SearchDataSourceFactory.kt
package com.pavelprymak.propodcast.presentation.paging

import androidx.paging.DataSource

import com.pavelprymak.propodcast.model.network.pojo.search.ResultsItem
import com.pavelprymak.propodcast.model.network.repo.PodcastRepoRx
import com.pavelprymak.propodcast.presentation.common.PagingStateBatch

class SearchDataSourceFactory : DataSource.Factory<Int, ResultsItem> {
    private val mPagingStateBatch: PagingStateBatch
    private val mRepoRx: PodcastRepoRx
    private val mSearchQuery: String
    private val mLanguage: String
    private var mPrevLoadingList: List<ResultsItem>? =null

    constructor(pagingStateBatch: PagingStateBatch, repoRx: PodcastRepoRx, searchQuery: String, language: String) {
        mPagingStateBatch = pagingStateBatch
        mRepoRx = repoRx
        mSearchQuery = searchQuery
        mLanguage = language
    }

    constructor(
        pagingStateBatch: PagingStateBatch,
        repoRx: PodcastRepoRx,
        searchQuery: String,
        language: String,
        prevLoadingList: List<ResultsItem>?
    ) {
        mPagingStateBatch = pagingStateBatch
        mRepoRx = repoRx
        mSearchQuery = searchQuery
        mLanguage = language
        mPrevLoadingList = prevLoadingList
    }

    override fun create(): DataSource<Int, ResultsItem> {
        return SearchDataSource(mRepoRx, mPagingStateBatch, mSearchQuery, mLanguage, mPrevLoadingList)
    }
}
